devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,366 @@
+
+    global_to_local_begin = Call(petsc_call, [Call('DMGlobalToLocalBegin', arguments=[
+        objs['da'], solver_objs['X_global'], 'INSERT_VALUES', solver_objs['X_local']])])
+
+    global_to_local_end = Call(petsc_call, [Call('DMGlobalToLocalEnd', arguments=[
+        objs['da'], solver_objs['X_global'], 'INSERT_VALUES', solver_objs['X_local']])])
+
+    dm_get_local_yvec = Call(petsc_call, [Call('DMGetLocalVector', arguments=[
+        objs['da'], Byref(solver_objs['Y_local'])])])
+
+    vec_get_array_y = Call(petsc_call, [Call('VecGetArray', arguments=[
+        solver_objs['Y_local'], Byref(petsc_arr_write.function)])])
+
+    vec_get_array_x = Call(petsc_call, [Call('VecGetArray', arguments=[
+        solver_objs['X_local'], Byref(petsc_arrays[0])])])
+
+    dm_get_local_info = Call(petsc_call, [Call('DMDAGetLocalInfo', arguments=[
+        objs['da'], Byref(petsc_arrays[0].function.dmda_info)])])
+
+    casts = [PointerCast(i.function) for i in petsc_arrays]
+
+    vec_restore_array_y = Call(petsc_call, [Call('VecRestoreArray', arguments=[
+        solver_objs['Y_local'], Byref(petsc_arr_write)])])
+
+    vec_restore_array_x = Call(petsc_call, [Call('VecRestoreArray', arguments=[
+        solver_objs['X_local'], Byref(petsc_arr_seed)])])
+
+    dm_local_to_global_begin = Call(petsc_call, [Call('DMLocalToGlobalBegin', arguments=[
+        objs['da'], solver_objs['Y_local'], 'INSERT_VALUES', solver_objs['Y_global']])])
+
+    dm_local_to_global_end = Call(petsc_call, [Call('DMLocalToGlobalEnd', arguments=[
+        objs['da'], solver_objs['Y_local'], 'INSERT_VALUES', solver_objs['Y_global']])])
+
+    func_return = Call('PetscFunctionReturn', arguments=[0])
+
+    matvec_body = List(body=[
+        petsc_func_begin_user, defn_struct, mat_get_dm, dm_get_app_context,
+        dm_get_local_xvec, global_to_local_begin, global_to_local_end,
+        dm_get_local_yvec, vec_get_array_y, vec_get_array_x, dm_get_local_info,
+        casts, BlankLine, body, vec_restore_array_y, vec_restore_array_x,
+        dm_local_to_global_begin, dm_local_to_global_end, func_return])
+
+    matvec_callback = Callable(
+        'MyMatShellMult_'+str(target.name), matvec_body, retval=objs['err'],
+        parameters=(solver_objs['Jac'], solver_objs['X_global'], solver_objs['Y_global']))
+
+    matvec_operation = Call(petsc_call, [
+        Call('MatShellSetOperation', arguments=[
+            solver_objs['Jac'], 'MATOP_MULT', Callback(matvec_callback.name,
+                                                       void, void)])])
+
+    return matvec_callback, matvec_operation
+
+
+def rebuild_expr_mapper(callable):
+
+    return {expr: expr._rebuild(
+        expr=expr.expr._rebuild(rhs=expr.expr.rhs.expr)) for
+        expr in FindNodes(LinearSolverExpression).visit(callable)}
+
+
+def transform_efuncs(efuncs, struct):
+
+    efuncs_new = []
+    for efunc in efuncs:
+        new_body = efunc.body
+        for i in struct.usr_ctx:
```

Comment:
Could you create the mapper `{i: FieldFromPointer(i, struct) for i in struct.usr_ctx}` and then just do a single `Uxreplace`?